compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -513,6 +513,57 @@ impl<'cx, 'tcx> TypeFolder<'tcx> for Canonicalizer<'cx, 'tcx> {
         let flags = FlagComputation::for_const(ct);
         if flags.intersects(self.needs_canonical_flags) { ct.super_fold_with(self) } else { ct }
     }
+
+    fn fold_effect(&mut self, e: ty::Effect<'tcx>) -> ty::Effect<'tcx> {
+        match e.val {
+            ty::EffectValue::Infer(ty::InferEffect::Var(vid)) => {
+                debug!("canonical: effect var found with vid {:?}", vid);
+                match self.infcx.probe_effect_var(vid) {
+                    Ok(e) => {
+                        debug!("(resolved to {:?})", e);
+                        return self.fold_effect(e);
+                    }
+
+                    // `InferEffect::Var(vid)` is unresolved, track its universe index in the
+                    // canonicalized result
+                    Err(mut ui) => {
+                        if !self.canonicalize_mode.preserve_universes() {
+                            // FIXME: perf problem described in #55921.
+                            ui = ty::UniverseIndex::ROOT;
+                        }
+                        return self.canonicalize_effect_var(
+                            CanonicalVarInfo {
+                                kind: CanonicalVarKind::Effect(ui, ty::EffectKind::Host),
+                            },
+                            e,
+                        );
+                    }
+                }
+            }
+            ty::EffectValue::Infer(ty::InferEffect::Fresh(_)) => {
+                bug!("encountered a fresh const during canonicalization")
+            }
+            ty::EffectValue::Bound(debruijn, _) => {
+                if debruijn >= self.binder_index {
+                    bug!("escaping bound effect during canonicalization")
+                } else {
+                    return e;
+                }
+            }
+            ty::EffectValue::Placeholder(placeholder) => {
+                return self.canonicalize_effect_var(
+                    CanonicalVarInfo {
+                        kind: CanonicalVarKind::PlaceholderEffect(placeholder, e.kind),
+                    },
+                    e,
+                );
+            }
+            _ => {}
+        }
+
+        let flags = FlagComputation::for_effect(e);
+        if flags.intersects(self.needs_canonical_flags) { e.super_fold_with(self) } else { e }
```

Comment:
why is that not simply happening in the `_ => {}` branch and we stop explicitly using `return` for the other branches.